### PR TITLE
Improve performance of parachain relay

### PR DESCRIPTION
**Pull request by vgeddes**
As part of the BEEFY/MMR proof generation, the parachain relay needs to find the relaychain block in which our parachain header was included (finalized). It used to do this by scanning all relaychain blocks backwards until it found a match. On our testnet we found that in certain scenarios, these scans often took hours or days to complete, causing denial of service.

It turns out that we can actually just query the parachain storage to find the relay chain block which _backed_ the parachain header. The relaychain block in which the parachain header was _included_ is usually 2-3 blocks after this:

image

Also refactored and cleaned up some of the relayer code for clarity. E2E tests are passing

Fixes: SNO-353

### CODE REVIEW

File: relayer/relays/parachain/scanner.go [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+// using the MMR root at the specified beefyBlockNumber of the relay chain.
+//
+// The algorithm works roughly like this:
+// 1. Fetch channel nonces on both sides of the bridge and compare them
+// 2. If the nonces on the parachain side are larger that means messages
+//    need to be relayed. If not then exit early.
+// 3. Scan parachain blocks to figure out exactly which commitments need to be relayed,
+// 4. For all the parachain blocks with unsettled commitments, determine the relay chain
+//    block number in which the parachain block was included,
+//
```
Comment (doubledup) [inline relayer/relays/parachain/scanner.go:44]: This looks incomplete 👀

**Review by alistair-singh — APPROVED** [review]

File: relayer/relays/parachain/scanner.go [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+// Scans for all parachain message commitments that need to be relayed and can be proven
+// using the MMR root at the specified beefyBlockNumber of the relay chain.
+//
+// The algorithm works roughly like this:
+// 1. Fetch channel nonces on both sides of the bridge and compare them
+// 2. If the nonces on the parachain side are larger that means messages
+//    need to be relayed. If not then exit early.
+// 3. Scan parachain blocks to figure out exactly which commitments need to be relayed,
+// 4. For all the parachain blocks with unsettled commitments, determine the relay chain
+//    block number in which the parachain block was included,
```
Comment (doubledup) [inline relayer/relays/parachain/scanner.go:43]:
```suggestion
// 3. Scan parachain blocks to figure out exactly which commitments need to be relayed.
// 4. For all the parachain blocks with unsettled commitments, determine the relay chain
//    block number in which the parachain block was included.
```

File: relayer/relays/parachain/scanner.go [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+//
+func (s *Scanner) Scan(ctx context.Context, beefyBlockNumber uint64) ([]*Task, error) {
+	beefyBlockHash, err := s.relayConn.API().RPC.Chain.GetBlockHash(uint64(beefyBlockNumber))
+	if err != nil {
+		return nil, fmt.Errorf("fetch block hash for block %v: %w", beefyBlockNumber, err)
+	}
+
+	// fetch last parachain header that was finalized *before* the BEEFY block
+	beefyBlockMinusOneHash, err := s.relayConn.API().RPC.Chain.GetBlockHash(uint64(beefyBlockNumber-1))
```
Comment (doubledup) [inline relayer/relays/parachain/scanner.go:52]: Why do we look at the second-last beefy block here? Are the parachain headers in the last block not finalized yet?

**Review by doubledup — APPROVED** [review]

File: relayer/relays/parachain/scanner.go (hunk 3 above)
Comment (vgeddes) [reply relayer/relays/parachain/scanner.go:52]: Its more that if a BEEFY commitment was generated for block X, then the MMR root hash in the commitment can only be used to prove parachain headers included before block X.

This explains it a bit more: https://docs.snowbridge.network/architecture/verification/polkadot#message-verification

File: relayer/relays/parachain/types.go [hunk 4]
```diff
@@ -9,15 +9,36 @@ import (
 	"github.com/snowfork/snowbridge/relayer/crypto/merkle"
 )
 
-// A Task is a bundle of items needed to submit commitments to Ethereum
+// A Task contains the working state for message commitments in a single parachain block
 type Task struct {
-	ParaID                        uint32
-	BlockNumber                   uint64
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply relayer/relays/parachain/types.go:15]: Yeah, it was redundant

File: relayer/relays/parachain/scanner.go [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("fetch block hash for block %v: %w", beefyBlockNumber, err)
+	}
+	var paraHead types.Header
+	ok, err := s.relayConn.FetchParachainHead(beefyBlockMinusOneHash, s.paraID, &paraHead)
+	if err != nil {
+		return nil, fmt.Errorf("fetch head for parachain %v at block %v: %w", s.paraID, beefyBlockHash.Hex(), err)
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply relayer/relays/parachain/scanner.go:59]: Nice catch!

**Pull request merged**